x/distribution/abci_test.go

```diff
@@ -0,0 +1,107 @@
+	app := simapp.Setup(t, false)
+	ctx := app.BaseApp.NewContext(false, tmproto.Header{})
+	addrs := simapp.AddTestAddrsIncremental(app, ctx, totalValidators, valTokens)
+	tstaking := teststaking.NewHelper(t, ctx, app.StakingKeeper)
+	tstaking.Commission = validatorCommissionRates
+
+	// create validators
+	validators := make([]validator, totalValidators-1)
+	for i := range validators {
+		validators[i].addr = sdk.ValAddress(addrs[i])
+		validators[i].pubkey = ed25519.GenPrivKey().PubKey()
+		validators[i].votes = make([]abci.VoteInfo, totalValidators)
+		tstaking.CreateValidatorWithValPower(validators[i].addr, validators[i].pubkey, power, true)
+	}
+	app.EndBlock(abci.RequestEndBlock{})
+	require.NotEmpty(t, app.Commit())
+
+	// verify validators lists
+	require.Len(t, app.StakingKeeper.GetAllValidators(ctx), totalValidators)
+	for i, val := range validators {
+		// verify all validator exists
+		require.NotNil(t, app.StakingKeeper.ValidatorByConsAddr(ctx, sdk.GetConsAddress(val.pubkey)))
+
+		// populate last commit info
+		voteInfos := []abci.VoteInfo{}
+		for _, val2 := range validators {
+			voteInfos = append(voteInfos, abci.VoteInfo{
+				Validator: abci.Validator{
+					Address: sdk.GetConsAddress(val2.pubkey),
+					Power:   power,
+				},
+				SignedLastBlock: true,
+			})
+		}
+
+		// have this validator only submit the minimum amount of pre-commits
+		if i == lazyValidatorIdx {
+			for j := totalValidators * 2 / 3; j < len(voteInfos); j++ {
+				voteInfos[j].SignedLastBlock = false
+			}
+		}
+
+		validators[i].votes = voteInfos
+	}
+
+	// previous block submitted by validator n-1 (with 100% previous commits) and proposed by lazy validator
+	app.BeginBlock(abci.RequestBeginBlock{
+		Header:         tmproto.Header{Height: app.LastBlockHeight() + 1, ProposerAddress: sdk.GetConsAddress(validators[lazyValidatorIdx].pubkey)},
+		LastCommitInfo: abci.LastCommitInfo{Votes: validators[lazyValidatorIdx-1].votes},
+	})
+	require.NotEmpty(t, app.Commit())
+
+	// previous block submitted by lazy validator (with 67% previous commits) and proposed by validator n+1
+	app.BeginBlock(abci.RequestBeginBlock{
+		Header:         tmproto.Header{Height: app.LastBlockHeight() + 1, ProposerAddress: sdk.GetConsAddress(validators[lazyValidatorIdx+1].pubkey)},
+		LastCommitInfo: abci.LastCommitInfo{Votes: validators[lazyValidatorIdx].votes},
+	})
+	require.NotEmpty(t, app.Commit())
+
+	// previous block submitted by validator n+1 (with 100% previous commits) and proposed by validator n+2
+	app.BeginBlock(abci.RequestBeginBlock{
+		Header:         tmproto.Header{Height: app.LastBlockHeight() + 1, ProposerAddress: sdk.GetConsAddress(validators[lazyValidatorIdx+2].pubkey)},
+		LastCommitInfo: abci.LastCommitInfo{Votes: validators[lazyValidatorIdx+1].votes},
+	})
+	require.NotEmpty(t, app.Commit())
+
+	rewardsValidatorBeforeLazyValidator := app.DistrKeeper.GetValidatorOutstandingRewardsCoins(ctx, validators[lazyValidatorIdx+1].addr)
```

Comment:
shouldnt this be a `-1`? 

otherwise `rewardsValidatorBeforeLazyValidator` and `rewardsValidatorAfterLazyValidator` are identical expressions so the `require.Equal` is a tautology

If i change it to `-1` the `require.Equal` fails. I'm not immediately sure why the rewards aren't equal but we might need a more sophisticated way to figure them out ...

Reply:
could it be because we never run a BeginBlock for BeforeLazyValidator?